### PR TITLE
Make underwater torches drop

### PR DESCRIPTION
**Pull request by ForbiddenJ**
Implements the suggestion #1830.

Torches drop when water floods over them.

I'm not sure how loud the sound should be, but I am certain it's quieter than a fire's equivalent.

### CODE REVIEW

**Comment by ForbiddenJ** [comment]
Travis CI says I left an unused variable in there? I guess I'll fix that when I learn of desired adjustments to the sound.

File: mods/default/torch.lua [hunk 1]
```diff
@@ -35,6 +35,12 @@ See LICENSE.txt and http://www.gnu.org/licenses/lgpl-2.1.txt
 
 --]]
 
+local on_flood = function(pos, oldnode, newnode)
+	local obj = minetest.add_item(pos, ItemStack("default:torch 1"))
```
Comment (sfan5) [inline mods/default/torch.lua]: unused var right here

**Comment by Fixer-007** [comment]
Is it water only or other liquids too (for example lava should flood torches too)?

**Comment by paramat** [comment]
I think in 'on_flood' check for newnode group = water otherwise return. Then lava will not drop the torch but will simply burn it as it should.

**Comment by ForbiddenJ** [comment]
Maybe to go even further, check if a liquid is part of the "igniter" group?

Minetest Game's [lava](https://github.com/minetest/minetest_game/blob/master/mods/default/nodes.lua#L1715) and Technic's [corium](https://github.com/minetest-mods/technic/blob/83a4bb60ca56a3e789aef4d91e747d2d6978eccd/technic/radiation.lua#L437) have this group in common... And they both burn up stuff!

**Comment by paramat** [comment]
If we allow lava to drop torches they become items that don't burn up.
But yes a check for igniter group is best, for liquids that should extinguish but are not in the water group.
So, return if the liquid is an igniter?

**Comment by paramat** [comment]
Doesn't seem right.
If liquid is an igniter, remove torch, no drop, nosound.
Otherwise drop torch with extinguish sound.

**Comment by ForbiddenJ** [comment]
That's what it does. The torch is removed regardless of the type of liquid, but no drop appears nor is there sound when the liquid is a part of the "igniter" group.

**Comment by paramat** [comment]
Sorry didn't see the 'not'.

**Comment by paramat** [comment]
Code seems good. If you've roughly balanced the sound against other game sounds that's fine, lavacooling volume is too loud for this certainly. We can always tune it later.

File: mods/default/torch.lua [hunk 2]
```diff
@@ -35,6 +35,19 @@ See LICENSE.txt and http://www.gnu.org/licenses/lgpl-2.1.txt
...
 
+local function on_flood(pos, oldnode, newnode)
+	local nodedef = minetest.registered_items[newnode.name]
+	-- Drop the torch if the liquid does not burn.
+	if nodedef ~= nil and not (
+			nodedef.groups ~= nil and
+			nodedef.groups.igniter ~= nil and
+			nodedef.groups.igniter > 0) then
+		minetest.add_item(pos, ItemStack("default:torch 1"))
+		minetest.sound_play("fire_extinguish_flame", {pos = pos, max_hear_distance = 8, gain = 0.07})
```
Comment (paramat) [inline mods/default/torch.lua]: Please split line before 80 columns (with tab size 4).

**Comment by paramat** [comment]
Otherwise +1.

File: mods/default/torch.lua (hunk 2 above)
Comment (ForbiddenJ) [inline mods/default/torch.lua]: Like this?

```
minetest.sound_play(
	"fire_extinguish_flame",
	{pos = pos, max_hear_distance = 8, gain = 0.07}
)
```

**Comment by paramat** [comment]
please could you make it
`minetest.sound_play("fire_extinguish_flame", {pos = pos, max_hear_distance = 16, gain = 0.1})`?
I find the distance too short and the sound too quiet.

**Comment by ForbiddenJ** [comment]
I added that to my adjustments. However, I find that it's almost as loud as a regular fire being extinguished.

**Comment by paramat** [comment]
Lavacooling is
```
minetest.sound_play("default_cool_lava",
{pos = pos, max_hear_distance = 16, gain = 0.25})
```

File: mods/default/torch.lua [hunk 3]
```diff
@@ -35,6 +35,27 @@ See LICENSE.txt and http://www.gnu.org/licenses/lgpl-2.1.txt
...
+	if nodedef == nil or not (
+			nodedef.groups ~= nil and
+			nodedef.groups.igniter ~= nil and
+			nodedef.groups.igniter > 0) then
+		minetest.add_item(pos, ItemStack("default:torch 1"))
+		minetest.sound_play(
+			"fire_extinguish_flame",
+			{pos = pos, max_hear_distance = 16, gain = 0.1}
+		)
+	else -- we're burning the torch.
```
Comment (paramat) [inline mods/default/torch.lua]: Please not this, fire_fire is too long a sample, and a torch submerged in lava would not be heard burning.

**Comment by paramat** [comment]
:+1: if you remove the torch burning code block.

**Comment by paramat** [comment]
Thanks :+1:

**Comment by Fixer-007** [comment]
Wait, you can still flood it with lava, dropped torch is destroyed by it (smoke as evidence, as any other wooden items).

**Comment by ForbiddenJ** [comment]
That is true. Sometime after I throw a torch into a lava pool, it burns up. I tested using bare Minetest Game.

**Comment by paramat** [comment]
Yeah i forgot that feature was merged.
Maybe if the liquid is an igniter we can run this code for a smoke puff and the 'burn-up' sound https://github.com/minetest/minetest_game/blob/master/mods/default/item_entity.lua#L16

**Comment by ForbiddenJ** [comment]
The bit that does the sound and smoke particle may need to be moved out as it looks a bit complex to be copied.

**Comment by paramat** [comment]
Yeah actually i'm beginning to think it's simpler to just drop the torch in lava and let the item burn up code do it's thing.

**Comment by ForbiddenJ** [comment]
So let the lava cause the torch to drop while still in absence of sound?

**Comment by paramat** [comment]
Yes, lava drops torch and leaves the burn up and sound to be done by the existing item burnup code.

**Comment by paramat** [comment]
:+1:

File: mods/default/torch.lua [hunk 4]
```diff
@@ -35,6 +35,26 @@ See LICENSE.txt and http://www.gnu.org/licenses/lgpl-2.1.txt
...
+local function on_flood(pos, oldnode, newnode)
+	-- Instantiate item drop.
+	minetest.add_item(pos, ItemStack("default:torch 1"))
+
+	-- Play sound if torch is not flooded by igniter type liquid.
+	local nodedef = minetest.registered_items[newnode.name]
+	if nodedef == nil or not (
+			nodedef.groups ~= nil and
+			nodedef.groups.igniter ~= nil and
+			nodedef.groups.igniter > 0) then
```
Comment (rubenwardy) [inline mods/default/torch.lua]:
```lua
if not (nodedef and nodedef.groups and nodedef.groups.igniter and nodedef.groups.igniter) then
```

equivalent under de Morgan’s Theorem and Double Negation.
`~= nil` is redundant as it's not a boolean value.
`a == nil` can be replaced with `not a`.

File: mods/default/torch.lua [hunk 5]
```diff
@@ -35,6 +35,26 @@ See LICENSE.txt and http://www.gnu.org/licenses/lgpl-2.1.txt
...
+		minetest.sound_play(
+			"default_cool_lava",
+			{pos = pos, max_hear_distance = 16, gain = 0.10}
+		)
```
Comment (rubenwardy) [inline mods/default/torch.lua]:
```lua
minetest.sound_play("default_cool_lava",
	{pos = pos, max_hear_distance = 16, gain = 0.10})
```

or 

```lua
minetest.sound_play("default_cool_lava", {
	pos = pos,
	max_hear_distance = 16,
	gain = 0.10
})
```

File: mods/default/torch.lua (hunk 4 above)
Comment (paramat) [inline mods/default/torch.lua]: Yeah i agree with this suggestion.

**Comment by paramat** [comment]
Tested with rubenwardy's logic suggestion.
Will edit that change in on merge.

**Comment by paramat** [comment]
3294a2a51549a32b9703e3a4d2346b7fe1b2756c

**Pull request closed without merging**